tokenizer/token.go

```diff
@@ -0,0 +1,596 @@
+}
+
+// Cause implements errors.Causer.
+func (e *TokenExtraError) Cause() error {
+	return e.Err
+}
+
+// Returns the ParseError object, if present.
+func (e *TokenExtraError) ParseError() *ParseError {
+	pe, ok := e.Err.(*ParseError)
+	if !ok {
+		return nil
+	}
+	return pe
+}
+
+func escapeIdentifier(s string) string { return escapeIdent(s, 0) }
+func escapeHashName(s string) string   { return escapeIdent(s, 1) }
+func escapeDimension(s string) string  { return escapeIdent(s, 2) }
+
+func needsHexEscaping(c byte, mode int) bool {
+	if c < 0x20 {
+		return true
+	}
+	if c >= utf8.RuneSelf {
+		return false
+	}
+	if mode == 2 {
+		if c == 'e' || c == 'E' {
+			return true
+		}
+	}
+	if c == '\\' {
+		return true
+	}
+	if isNameCode(c) {
+		return false
+	}
+	return true
+}
+
+func escapeIdent(s string, mode int) string {
+	if s == "" {
+		return ""
+	}
+	var buf bytes.Buffer
+	buf.Grow(len(s))
+	anyChanges := false
+
+	var i int
+
+	// Handle first character
+	// dashes allowed at start only for TokenIdent-ish
+	// eE not allowed at start for Dimension
+	if mode != 1 {
+		if !isNameStart(s[0]) && s[0] != '-' && s[0] != 'e' && s[0] != 'E' {
+			if needsHexEscaping(s[0], mode) {
+				fmt.Fprintf(&buf, "\\%X ", s[0])
+				anyChanges = true
+			} else {
+				buf.WriteByte('\\')
+				buf.WriteByte(s[0])
+				anyChanges = true
+			}
+		} else if s[0] == 'e' || s[0] == 'E' {
+			if mode == 2 {
+				fmt.Fprintf(&buf, "\\%X ", s[0])
+				anyChanges = true
+			} else {
+				buf.WriteByte(s[0])
+			}
+		} else if s[0] == '-' {
+			if len(s) == 1 {
+				return "\\-"
+			} else if isNameStart(s[1]) {
+				buf.WriteByte('-')
+			} else {
+				buf.WriteString("\\-")
+				anyChanges = true
+			}
+		} else {
+			buf.WriteByte(s[0])
+		}
+		i = 1
+	} else {
+		i = 0
+	}
+	// Write the rest of the name
+	for ; i < len(s); i++ {
+		if !isNameCode(s[i]) {
+			fmt.Fprintf(&buf, "\\%X ", s[i])
+			anyChanges = true
+		} else {
+			buf.WriteByte(s[i])
+		}
+	}
+
+	if !anyChanges {
+		return s
+	}
+	return buf.String()
+}
+
+func escapeString(s string, delim byte) string {
+	var buf bytes.Buffer
+	if delim != 0 {
+		buf.WriteByte(delim)
+	}
+	for i := 0; i < len(s); i++ {
+		switch s[i] {
+		case '"':
+			buf.WriteString("\\\"")
+			continue
+		case delim:
+			buf.WriteByte('\\')
+			buf.WriteByte(delim)
+			continue
+		case '\n':
+			buf.WriteString("\\0A ")
+			continue
+		case '\r':
+			buf.WriteString("\\0D ")
+			continue
+		case '\\':
+			buf.WriteString("\\\\")
+			continue
+		}
+		if s[i] < utf8.RuneSelf && isNonPrintable(s[i]) {
+			fmt.Fprintf(&buf, "\\%X ", s[i])
+			continue
+		}
+		buf.WriteByte(s[i])
+	}
+	if delim != 0 {
+		buf.WriteByte(delim)
+	}
+	return buf.String()
+}
+
+// Attempt to turn the token back into a CSS string.  (Wrapper around WriteTo.)
+func (t *Token) Render() string {
+	var buf bytes.Buffer
+	t.WriteTo(&buf)
+	return buf.String()
+}
+
+// Attempt to turn the token back into a CSS string.
+func (t *Token) WriteTo(w io.Writer) {
```

Comment:
This is causing `go vet` to raise an error in the Travis tests. The signature should be `func (t *Token) WriteTo(w io.Writer) (int64, error)` with the same kind of semantics as the interface (return the number of bytes written and any error).

Reply:
Yeah, I figured that was going to happen. Do you think I should just do Render() string or implement the byte counting?

Reply:
At the least it needs to return the error from the underlying writer, so you might as well include the byte counting.

Reply:
It should be pretty easy if you eliminate the uses of `fmt` and just use `io.WriteString` everywhere

Reply:
Ended up using a helper function, called it stickyWriteString(n *int64, err *error, w io.Writer, s string) after the concept of "sticky errors". Not the best name but /shrug

Also updated the README and did another pass over documentation.